proxy.go

```diff
@@ -0,0 +1,128 @@
+package main
+
+import (
+	"errors"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// Log forwarders like fluentd or fluentbit post log lines in bulk to this
+// elasticsearch endpoint
+const BulkLogsEndpoint = "/_bulk"
+
+func unauthorized(w http.ResponseWriter) {
+	w.WriteHeader(http.StatusUnauthorized)
+	w.Write([]byte("{\"message\":\"unauthorized\"}"))
+}
+
+func copyHeader(dst, src http.Header) {
+	for k, vv := range src {
+		for _, v := range vv {
+			dst.Add(k, v)
+		}
+	}
+}
+
+type Proxy struct {
+	config        *ProxyConfig
+	httpClient    *http.Client
+	remoteURL     *url.URL
+	signer        Signer
+	authenticator Authenticator
+}
+
+// isValidTargetEndpoint checks if the request has a valid target depending
+// on the configuration of the proxy
+func (p *Proxy) isValidTargetEndpoint(r *http.Request) bool {
+	return !p.config.LogsOnly || (p.config.LogsOnly && r.URL.Path == BulkLogsEndpoint)
+
+}
+
+func (p *Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if !p.isValidTargetEndpoint(r) {
+		unauthorized(w)
+		return
+	}
+
+	if p.config.InputValidation {
+		err := p.authenticator.AuthenticateRequest(r)
+		if err != nil {
+			msg := "not authorized to proxy request"
+			http.Error(w, msg, http.StatusUnauthorized)
```

Comment:
Are we sure we want to return a blanket "not authorized to proxy request" error to the caller for _all_ errors? As we are integrating with the proxy, we might mess up and forget a header, use an incorrect ID, etc... With this blanket error returned, we'd have to check the proxy logs to understand what we did wrong. I would suggest that we create custom Error structs for the two authorization-related errors returned from `AuthenticateRequest` (i.e., `ErrIDNotWhitelisted` and `ErrSignatureInvalid`):

```go
type ErrIDNotWhitelisted struct {
  ID string
}

func (e ErrIDNotWhitelisted) Error() string {
    return fmt.Sprintf("provided Proxy-ID is not whitelisted: %s", e.ID)
}

type ErrSignatureInvalid struct {}

func (e ErrSignatureInvalid) Error() string {
    return "Request signature invalid"
}
```

And then after checking `err != nil`, we can check for these errors and for all other errors return a 400 (bad request) with a helpful error message:

```
if err != nil {
    switch err.(type) {
	case ErrIDNotWhitelisted:
        case ErrSignatureInvalid:
		msg := "not authorized to proxy request"
		http.Error(w, msg, http.StatusUnauthorized)
         default:
	          http.Error(w, err.Error(), http.StatusBadRequest)
    }
    log.WithField("error", err).Error("AuthenticateRequest returned an error")
}
```